File: apps/api/app/services/lyric_search.py
from __future__ import annotations

import re
import unicodedata
from dataclasses import dataclass
from functools import lru_cache

from app.models import Song
from app.schemas.search import MediaSummary, SearchResponse, SearchResultItem
from app.services.catalog import catalog_song_snapshot

LYRIC_RESULT_LIMIT = 5
LYRIC_PHRASE_SCORE = 48.0
LYRIC_ENGLISH_SCORE = 40.0
_TOKEN = re.compile(r"[^a-z0-9]+")
_STOP = frozenset(
    (
        "a an and as at be but by can do for from i if in is it me my no not o of oh "
        "on or so that the this to us we with you your"
    ).split()
)


def normalize_lyric_text(value: str | None) -> str:
    decomposed = unicodedata.normalize("NFKD", value or "").casefold()
    plain = "".join(character for character in decomposed if not unicodedata.combining(character))
    return _TOKEN.sub(" ", plain).strip()


def fold_lyric_phonetic(value: str) -> str:
    """Fold common Roman-transliteration spellings: humdardi ≈ hamdardi, siv ≈ shiva."""
    return " ".join(token for token in (_fold_lyric_token(part) for part in value.split()) if token)


def _fold_lyric_token(token: str) -> str:
    folded = token
    replacements = (
        (r"aa+", "a"),
        (r"ee+", "i"),
        (r"oo+", "u"),
        (r"uu+", "u"),
        (r"kh", "k"),
        (r"gh", "g"),
        (r"bh", "b"),
        (r"dh", "d"),
        (r"ph", "f"),
        (r"th", "t"),
        (r"sh", "s"),
        (r"ch", "c"),
        (r"v", "w"),
        (r"y", "i"),
    )
    for pattern, replacement in replacements:
        folded = re.sub(pattern, replacement, folded)
    if len(folded) >= 4 and folded.endswith("a"):
        folded = folded[:-1]
    folded = re.sub(r"[ou]", "a", folded)
    folded = re.sub(r"i{2,}", "i", folded)
    folded = re.sub(r"a{2,}", "a", folded)
    return folded


def max_lyric_edits(length: int) -> int:
    if length < 4:
        return 0
    if length < 6:
        return 1
    return 2


def _adjacent_transpose(left: str, right: str) -> bool:
    if len(left) != len(right) or len(left) < 2:
        return False
    index = 0
    while index < len(left) and left[index] == right[index]:
        index += 1
    if index >= len(left) - 1:
        return False
    if left[index] != right[index + 1] or left[index + 1] != right[index]:
        return False
    return left[index + 2 :] == right[index + 2 :]


def within_lyric_edits(left: str, right: str, max_edits: int | None = None) -> bool:
    allowed = max_lyric_edits(len(left)) if max_edits is None else max_edits
    if left == right:
        return True
    if allowed <= 0:
        return False
    if abs(len(left) - len(right)) > allowed:
        return False
    if len(left) == len(right) and _adjacent_transpose(left, right):
        return True
    previous = list(range(len(right) + 1))
    current = [0] * (len(right) + 1)
    for row, left_ch in enumerate(left, start=1):
        current[0] = row
        best = current[0]
        for col, right_ch in enumerate(right, start=1):
            cost = 0 if left_ch == right_ch else 1
            value = min(previous[col] + 1, current[col - 1] + 1, previous[col - 1] + cost)
            current[col] = value
            if value < best:
                best = value
        if best > allowed:
            return False
        previous, current = current, previous
    return previous[len(right)] <= allowed


def fuzzy_token_match(needle: str, tokens: tuple[str, ...] | frozenset[str]) -> bool:
    if not needle or len(needle) < 4:
        return False
    allowed = max_lyric_edits(len(needle))
    if allowed <= 0:
        return False
    for token in tokens:
        if not token or len(token) < 3:
            continue
        if token == needle:
            return True
        if token.startswith(needle) and len(token) - len(needle) <= allowed + 1:
            return True
        if needle.startswith(token) and len(needle) - len(token) <= allowed + 1:
            return True
        if abs(len(token) - len(needle)) > allowed:
            continue
        if within_lyric_edits(needle, token, allowed):
            return True
    return False


def lyric_tokens_match(left: str, right: str) -> bool:
    if not left or not right:
        return False
    if left == right:
        return True
    a = fold_lyric_phonetic(left)
    b = fold_lyric_phonetic(right)
    if not a or not b:
        return False
    if a == b:
        return True
    if len(a) >= 5 and len(b) > len(a) and b.startswith(a) and len(b) - len(a) <= 2:
        return True
    if len(b) >= 5 and len(a) > len(b) and a.startswith(b) and len(a) - len(b) <= 2:
        return True
    if len(a) < 5 or len(b) < 5:
        return False
    if a[:4] != b[:4]:
        return False
    return within_lyric_edits(a, b, 1)


def ordered_lyric_coverage(query_tokens: list[str], haystack_tokens: list[str]) -> float:
    if not query_tokens or not haystack_tokens:
        return 0.0
    best = 0.0
    max_gap = 1 if len(query_tokens) >= 4 else 2
    first = query_tokens[0]
    for start, hay in enumerate(haystack_tokens):
        if not lyric_tokens_match(first, hay):
            continue
        qi = 1
        gaps = 0
        for hi in range(start + 1, len(haystack_tokens)):
            if qi >= len(query_tokens):
                break
            if lyric_tokens_match(query_tokens[qi], haystack_tokens[hi]):
                qi += 1
                gaps = 0
                continue
            gaps += 1
            if gaps > max_gap:
                break
        best = max(best, qi / len(query_tokens))
        if best == 1:
            return 1.0
    return best


def _has_lyric_phrase_anchor(
    anchors: list[str],
    record: LyricRecord,
    opening_words: list[str],
    body_words: list[str],
) -> bool:
    for anchor in anchors:
        if len(anchor) < 4 or anchor in _COMMON_LYRIC_TOKENS:
            continue
        if anchor in record.folded_tokens or anchor in record.opening_token_set:
            return True
        if any(lyric_tokens_match(anchor, token) for token in opening_words):
            return True
        if any(lyric_tokens_match(anchor, token) for token in body_words):
            return True
    return False


_COMMON_LYRIC_TOKENS = frozenset(
    (
        "ami tumi tomar tomay tomake tomakei mora mor mama go re se oi ei ar na ki kii he "
        "ogo prabhu more moreke amay amake"
    ).split()
)


@dataclass(frozen=True, slots=True)
class LyricHit:
    number: int
    score: float
    matched_by: str


@dataclass(frozen=True, slots=True)
class LyricRecord:
    number: int
    title: str
    opening: str
    body: str
    tokens: frozenset[str]
    folded_title: str
    folded_opening: str
    folded_body: str
    folded_tokens: frozenset[str]
    opening_tokens: tuple[str, ...]
    opening_token_set: frozenset[str]
    token_list: tuple[str, ...]
    opening_raw_tokens: tuple[str, ...]
    raw_token_list: tuple[str, ...]


def _record_from_song(song: Song) -> LyricRecord:
    title = normalize_lyric_text(song.title)
    opening = normalize_lyric_text(song.first_line)
    lyrics = normalize_lyric_text(song.lyrics_original)
    translit = normalize_lyric_text(song.transliteration)
    body = " ".join(part for part in (title, opening, lyrics, translit) if part)
    folded_title = fold_lyric_phonetic(title)
    folded_opening = fold_lyric_phonetic(opening)
    folded_body = fold_lyric_phonetic(body)
    folded_tokens = frozenset(folded_body.split())
    opening_tokens = tuple(
        dict.fromkeys(
            token
            for token in f"{folded_title} {folded_opening}".split()
            if len(token) >= 3
        )
    )
    opening_raw_tokens = tuple(
        dict.fromkeys(
            token
            for token in f"{title} {opening}".split()
            if len(token) >= 3
        )
    )
    raw_token_list = tuple(dict.fromkeys(token for token in body.split() if len(token) >= 3))
    return LyricRecord(
        number=int(song.number),
        title=title,
        opening=opening,
        body=body,
        tokens=frozenset(body.split()),
        folded_title=folded_title,
        folded_opening=folded_opening,
        folded_body=folded_body,
        folded_tokens=folded_tokens,
        opening_tokens=opening_tokens,
        opening_token_set=frozenset(opening_tokens),
        token_list=tuple(folded_tokens),
        opening_raw_tokens=opening_raw_tokens,
        raw_token_list=raw_token_list,
    )


@lru_cache(maxsize=1)
def lyric_index() -> tuple[tuple[LyricRecord, ...], dict[str, frozenset[int]]]:
    records = tuple(_record_from_song(song) for song in catalog_song_snapshot())
    postings: dict[str, set[int]] = {}
    for index, record in enumerate(records):
        for token in record.tokens:
            bucket = postings.get(token)
            if bucket is None:
                postings[token] = {index}
            else:
                bucket.add(index)
    frozen = {token: frozenset(rows) for token, rows in postings.items()}
    return records, frozen


def _candidate_indexes(tokens: list[str], postings: dict[str, frozenset[int]]) -> set[int]:
    lists = [postings[token] for token in tokens if token in postings]
    if not lists:
        return set()
    lists.sort(key=len)
    matched = set(lists[0])
    for bucket in lists[1:]:
        matched &= bucket
        if not matched:
            break
    if matched:
        return matched
    return set(lists[0])


def _token_in_record(token: str, folded_token: str, record: LyricRecord) -> bool:
    if token in record.tokens:
        return True
    if len(token) < 4:
        return False
    return folded_token in record.folded_tokens


def _score_record(
    query: str,
    tokens: list[str],
    record: LyricRecord,
    folded_query: str,
    folded_tokens: list[str],
) -> LyricHit | None:
    if query == record.opening or query == record.title:
        return LyricHit(record.number, 100.0, "opening_line")
    if record.opening.startswith(query) or record.title.startswith(query):
        return LyricHit(record.number, 88.0, "opening_line")
    if query in record.opening or query in record.title:
        return LyricHit(record.number, 72.0, "opening_line")
    if query in record.body:
        return LyricHit(record.number, 48.0, "full_text")
    if len(folded_query) >= 4 and (
        folded_query in record.folded_opening or folded_query in record.folded_title
    ):
        return LyricHit(record.number, 64.0, "opening_line")
    if len(folded_query) >= 3 and folded_query in record.opening_token_set:
        return LyricHit(record.number, 64.0, "opening_line")
    if len(folded_query) >= 4 and folded_query in record.folded_body:
        return LyricHit(record.number, 44.0, "full_text")
    if len(folded_query) >= 3 and folded_query in record.folded_tokens:
        return LyricHit(record.number, 44.0, "full_text")

    if len(tokens) >= 3:
        opening_words = f"{record.title} {record.opening}".split()
        body_words = record.body.split()
        folded_opening_words = f"{record.folded_title} {record.folded_opening}".split()
        if _has_lyric_phrase_anchor(folded_tokens, record, opening_words, body_words):
            opening_coverage = max(
                ordered_lyric_coverage(tokens, opening_words),
                ordered_lyric_coverage(folded_tokens, folded_opening_words),
            )
            body_coverage = max(
                ordered_lyric_coverage(tokens, body_words),
                ordered_lyric_coverage(folded_tokens, record.folded_body.split()),
            )
            best_coverage = max(opening_coverage, body_coverage)
            if best_coverage >= 0.8:
                score = round(70 + best_coverage * 18)
                matched_by = "opening_line" if opening_coverage >= body_coverage else "full_text"
                return LyricHit(record.number, float(score), matched_by)

    if not tokens:
        return None
    distinctive = [token for token in tokens if token not in _STOP]
    scored = distinctive or tokens
    folded_map = dict(zip(tokens, folded_tokens, strict=True))
    opening_hits = 0
    body_hits = 0
    unmatched: list[tuple[str, str]] = []
    for token in scored:
        folded_token = folded_map[token]
        if token in record.opening_raw_tokens or (
            len(folded_token) >= 3 and folded_token in record.opening_token_set
        ):
            opening_hits += 1
            continue
        if token in record.tokens or (
            len(folded_token) >= 3 and folded_token in record.folded_tokens
        ):
            body_hits += 1
            continue
        unmatched.append((token, folded_token))
    matched = opening_hits + body_hits
    if not (matched / len(scored) >= 0.6) and unmatched and (len(scored) <= 2 or matched > 0):
        for token, folded_token in unmatched:
            if fuzzy_token_match(token, record.opening_raw_tokens) or fuzzy_token_match(
                folded_token, record.opening_tokens
            ):
                opening_hits += 1
                continue
            if fuzzy_token_match(token, record.raw_token_list) or fuzzy_token_match(
                folded_token, record.token_list
            ):
                body_hits += 1
    coverage = (opening_hits + body_hits) / len(scored)
    rare_opening_hits = 0
    for token in scored:
        folded_token = folded_map[token]
        if token in _COMMON_LYRIC_TOKENS or folded_token in _COMMON_LYRIC_TOKENS:
            continue
        if (
            token in record.opening_raw_tokens
            or (len(folded_token) >= 3 and folded_token in record.opening_token_set)
            or fuzzy_token_match(token, record.opening_raw_tokens)
            or fuzzy_token_match(folded_token, record.opening_tokens)
        ):
            rare_opening_hits += 1
    if opening_hits and opening_hits / len(scored) >= 0.6:
        if len(scored) >= 3 and rare_opening_hits == 0:
            return LyricHit(record.number, 36.0, "opening_line")
        if len(scored) == 1:
            return LyricHit(record.number, 64.0, "opening_line")
        if len(scored) >= 3:
            return LyricHit(record.number, 52.0, "opening_line")
        return LyricHit(record.number, 58.0, "opening_line")
    if opening_hits + body_hits and coverage >= 0.6:
        return LyricHit(record.number, 44.0 if len(scored) == 1 else 42.0, "full_text")
    return None


def search_lyrics(query: str, limit: int = LYRIC_RESULT_LIMIT) -> list[LyricHit]:
    normalized = normalize_lyric_text(query)
    if len(normalized) < 2:
        return []
    records, _postings = lyric_index()
    tokens = normalized.split()
    folded_query = fold_lyric_phonetic(normalized)
    folded_tokens = [fold_lyric_phonetic(token) for token in tokens]
    hits: list[LyricHit] = []
    for record in records:
        hit = _score_record(normalized, tokens, record, folded_query, folded_tokens)
        if hit:
            hits.append(hit)
    hits.sort(key=lambda item: (-item.score, item.number))
    return hits[: max(1, limit)]


def confident_lyric_hits(hits: list[LyricHit]) -> list[LyricHit]:
    return [
        hit
        for hit in hits
        if hit.matched_by == "opening_line" or hit.score >= LYRIC_ENGLISH_SCORE
    ]


def lyric_search_response(
    query: str,
    hits: list[LyricHit],
    songs: list[Song],
    media_counts: dict[int, MediaSummary],
) -> SearchResponse:
    lookup = {song.number: song for song in songs}
    items: list[SearchResultItem] = []
    for hit in hits:
        song = lookup.get(hit.number)
        if song is None:
            continue
        summary = media_counts.get(song.number, MediaSummary())
        matched_by = [hit.matched_by]
        if song.is_verified or song.canonical_source_status == "verified":
            matched_by.append("verified")
        items.append(
            SearchResultItem(
                song_number=song.number,
                opening_line=song.first_line,
                matched_by=matched_by,
                score=round(hit.score, 4),
                verification_status="officially_verified"
                if song.is_verified or song.canonical_source_status == "verified"
                else song.canonical_source_status,
                themes=[value for value in [song.theme] if value],
                media_summary=summary,
            )
        )
    return SearchResponse(
        query=query,
        detected_intent="lyric_search",
        total=len(items),
        items=items,
    )

File: packages/core/src/lyric-search.ts
export type LyricSearchRow = {
  n: number
  t: string
  o: string
  b: string
}

export type LyricSearchHit = {
  number: number
  title: string
  firstLine: string
  snippet: string
  score: number
  matchedBy: "opening_line" | "full_text"
}

const TOKEN = /[^a-z0-9]+/g

const STOP = new Set(
  "a an and as at be but by can do for from i if in is it me my no not o of oh on or so that the this to us we with you your".split(
    " ",
  ),
)

export function isCatalogNumberQuery(value: string) {
  return /^\s*(?:ps[\s-]*)?\d{1,4}\s*$/i.test(value)
}

export function normalizeLyricText(value: string | null | undefined): string {
  const decomposed = (value ?? "").normalize("NFKD").toLowerCase()
  let plain = ""
  for (const character of decomposed) {
    if (character.charCodeAt(0) >= 0x300 && character.charCodeAt(0) <= 0x36f) continue
    plain += character
  }
  return plain.replace(TOKEN, " ").trim()
}

/** Fold common Roman-transliteration spellings: humdardi ≈ hamdardi, siv ≈ shiva. */
export function foldLyricPhonetic(value: string): string {
  return value
    .split(" ")
    .map((token) => foldLyricToken(token))
    .filter(Boolean)
    .join(" ")
}

function foldLyricToken(token: string): string {
  if (!token) return token
  let folded = token
  folded = folded.replace(/aa+/g, "a")
  folded = folded.replace(/ee+/g, "i")
  folded = folded.replace(/oo+/g, "u")
  folded = folded.replace(/uu+/g, "u")
  folded = folded.replace(/kh/g, "k")
  folded = folded.replace(/gh/g, "g")
  folded = folded.replace(/bh/g, "b")
  folded = folded.replace(/dh/g, "d")
  folded = folded.replace(/ph/g, "f")
  folded = folded.replace(/th/g, "t")
  folded = folded.replace(/sh/g, "s")
  folded = folded.replace(/ch/g, "c")
  folded = folded.replace(/v/g, "w")
  folded = folded.replace(/y/g, "i")
  if (folded.length >= 4 && folded.endsWith("a")) folded = folded.slice(0, -1)
  folded = folded.replace(/[ou]/g, "a")
  folded = folded.replace(/i{2,}/g, "i")
  folded = folded.replace(/a{2,}/g, "a")
  return folded
}

/** Names people type many ways: siv / shiv / shiva, krishna / krsna. */
const SEARCH_NAME_ALIASES: Record<string, string> = {
  siv: "shiva",
  shiv: "shiva",
  siva: "shiva",
  shiva: "shiva",
  mahadev: "shiva",
  mahadeva: "shiva",
  shankar: "shiva",
  shankara: "shiva",
  sadashiv: "shiva",
  sadashiva: "shiva",
  kishan: "krsna",
  kisan: "krsna",
  kisna: "krsna",
  kisn: "krsna",
  kishna: "krsna",
  krisna: "krsna",
  krisn: "krsna",
  krishn: "krsna",
  krishna: "krsna",
  krsna: "krsna",
  krsn: "krsna",
  krshna: "krsna",
  krushna: "krsna",
  krushn: "krsna",
}

export function canonicalSearchKey(value: string): string {
  const normalized = normalizeLyricText(value)
  if (!normalized) return normalized
  if (SEARCH_NAME_ALIASES[normalized]) return SEARCH_NAME_ALIASES[normalized]
  const folded = foldLyricPhonetic(normalized)
  return SEARCH_NAME_ALIASES[folded] ?? SEARCH_NAME_ALIASES[normalized.replace(/a$/, "")] ?? normalized
}

/** Drop “songs of / songs about / …songs” so any remaining word can match the catalog. */
export function stripCatalogSearchFraming(value: string): string {
  return normalizeLyricText(value)
    .replace(/^(search prabhat samgiita for)\s+/, "")
    .replace(/^(songs?)\s+(of|for|about|on)\s+/, "")
    .replace(/\s+(songs?|music|bhajans?|kirtans?|kiirtans?)$/, "")
    .trim()
}

/** Keys to try against collections or lyrics: full query, stripped framing, short-name tokens. */
export function catalogLookupKeys(query: string): string[] {
  const keys: string[] = []
  const add = (value: string) => {
    if (!value || keys.includes(value)) return
    keys.push(value)
    const compact = value.replace(/ /g, "")
    if (compact !== value) add(compact)
    const canon = canonicalSearchKey(value)
    if (canon !== value) add(canon)
  }
  const full = normalizeLyricText(query)
  add(full)
  const stripped = stripCatalogSearchFraming(query)
  add(stripped)
  const words = stripped.split(" ").filter(Boolean)
  if (words.length >= 1 && words.length <= 3) {
    for (const word of words) {
      if (word.length >= 3 && !STOP.has(word)) add(word)
    }
  }
  return keys
}

/** One typo for short words, two for longer ones. */
export function maxLyricEdits(length: number) {
  if (length < 5) return 0
  if (length < 8) return 1
  return 2
}

function adjacentTranspose(left: string, right: string) {
  if (left.length !== right.length || left.length < 2) return false
  let index = 0
  while (index < left.length && left[index] === right[index]) index += 1
  if (index >= left.length - 1) return false
  if (left[index] !== right[index + 1] || left[index + 1] !== right[index]) return false
  return left.slice(index + 2) === right.slice(index + 2)
}

export function withinLyricEdits(left: string, right: string, max = maxLyricEdits(left.length)) {
  if (left === right) return true
  if (max <= 0) return false
  const gap = Math.abs(left.length - right.length)
  if (gap > max) return false
  if (gap === 0 && adjacentTranspose(left, right)) return true
  const rows = left.length
  const cols = right.length
  let previous = new Array(cols + 1)
  let current = new Array(cols + 1)
  for (let col = 0; col <= cols; col += 1) previous[col] = col
  for (let row = 1; row <= rows; row += 1) {
    current[0] = row
    let best = current[0]
    const leftCode = left.charCodeAt(row - 1)
    for (let col = 1; col <= cols; col += 1) {
      const cost = leftCode === right.charCodeAt(col - 1) ? 0 : 1
      const value = Math.min(previous[col] + 1, current[col - 1] + 1, previous[col - 1] + cost)
      current[col] = value
      if (value < best) best = value
    }
    if (best > max) return false
    const swap = previous
    previous = current
    current = swap
  }
  return previous[cols] <= max
}

function fuzzyTokenMatch(needle: string, tokens: Iterable<string>) {
  const max = maxLyricEdits(needle.length)
  if (max <= 0 || needle.length < 2) return false
  const first = needle[0]
  const second = needle[1]
  for (const token of tokens) {
    if (!token) continue
    if (token[0] !== first && token[0] !== second) continue
    if (token === needle) return true
    if (needle.length >= 5 && token.startsWith(needle) && token.length - needle.length <= max) return true
    if (token.length >= 5 && needle.startsWith(token) && needle.length - token.length <= max) return true
    if (withinLyricEdits(needle, token, max)) return true
  }
  return false
}

function lyricSnippet(query: string, tokens: string[], opening: string, body: string) {
  const source = body.includes(query) ? body : opening
  const at = source.indexOf(query)
  if (at >= 0) {
    const start = Math.max(0, at - 20)
    const end = Math.min(source.length, at + Math.max(query.length, 48) + 28)
    let slice = source.slice(start, end).trim()
    if (start > 0) slice = `…${slice}`
    if (end < source.length) slice = `${slice}…`
    return slice
  }
  const distinctive = tokens.filter((token) => !STOP.has(token) && body.includes(token))
  if (distinctive.length) {
    const first = body.indexOf(distinctive[0])
    const start = Math.max(0, first - 16)
    const end = Math.min(body.length, first + 72)
    let slice = body.slice(start, end).trim()
    if (start > 0) slice = `…${slice}`
    if (end < body.length) slice = `${slice}…`
    return slice
  }
  const foldedQuery = foldLyricPhonetic(query)
  if (foldedQuery.length >= 4) {
    const words = body.split(" ").filter(Boolean)
    const needle = foldedQuery.split(" ")[0] ?? ""
    const index = words.findIndex((word) => {
      const folded = foldLyricPhonetic(word)
      return folded.includes(needle) || (needle.length >= 5 && withinLyricEdits(needle, folded))
    })
    if (index >= 0) {
      const start = Math.max(0, index - 2)
      const end = Math.min(words.length, index + 8)
      return words.slice(start, end).join(" ")
    }
  }
  return opening
}

type FoldedLyric = {
  title: string
  opening: string
  body: string
  tokens: Set<string>
  openingTokens: string[]
  tokenList: string[]
}

const foldedCache = new WeakMap<LyricSearchRow, FoldedLyric>()

function foldedLyric(row: LyricSearchRow): FoldedLyric {
  const cached = foldedCache.get(row)
  if (cached) return cached
  const title = foldLyricPhonetic(normalizeLyricText(row.t))
  const opening = foldLyricPhonetic(normalizeLyricText(row.o))
  const body = foldLyricPhonetic(row.b)
  const tokens = new Set(body.split(" ").filter(Boolean))
  const openingTokens = Array.from(
    new Set(`${title} ${opening}`.split(" ").filter((token) => token.length >= 4)),
  )
  const next = {
    title,
    opening,
    body,
    tokens,
    openingTokens,
    tokenList: Array.from(tokens),
  }
  foldedCache.set(row, next)
  return next
}

function scoreRow(
  query: string,
  tokens: string[],
  row: LyricSearchRow,
  foldedQuery: string,
  foldedTokenByExact: Map<string, string>,
): LyricSearchHit | null {
  const title = normalizeLyricText(row.t)
  const opening = normalizeLyricText(row.o)
  const body = row.b
  const bodyTokens = new Set(body.split(" ").filter(Boolean))
  const folded = foldedLyric(row)
  let score = 0
  let matchedBy: LyricSearchHit["matchedBy"] = "full_text"
  if (query === opening || query === title) {
    score = 100
    matchedBy = "opening_line"
  } else if (opening.startsWith(query) || title.startsWith(query)) {
    score = 88
    matchedBy = "opening_line"
  } else if (opening.includes(query) || title.includes(query)) {
    score = 72
    matchedBy = "opening_line"
  } else if (body.includes(query)) {
    score = 48
    matchedBy = "full_text"
  } else if (foldedQuery.length >= 4 && (folded.opening.includes(foldedQuery) || folded.title.includes(foldedQuery))) {
    score = 64
    matchedBy = "opening_line"
  } else if (foldedQuery.length >= 4 && folded.body.includes(foldedQuery)) {
    score = 44
    matchedBy = "full_text"
  } else if (tokens.length) {
    const distinctive = tokens.filter((token) => !STOP.has(token))
    const scored = distinctive.length ? distinctive : tokens
    const exactHits = scored.filter((token) => {
      if (bodyTokens.has(token)) return true
      if (token.length < 4) return false
      const foldedToken = foldedTokenByExact.get(token)
      return Boolean(foldedToken) && folded.tokens.has(foldedToken)
    }).length
    const coverage = exactHits / scored.length
    if (exactHits && distinctive.length >= 3 && coverage >= 0.7) {
      score = 40
      matchedBy = "full_text"
    } else {
      const needles = scored
        .map((token) => foldedTokenByExact.get(token) ?? foldLyricPhonetic(token))
        .filter((token) => token.length >= 5)
      if (needles.length) {
        const openingHits = needles.filter((needle) => fuzzyTokenMatch(needle, folded.openingTokens)).length
        if (openingHits / needles.length >= 0.6) {
          score = 58
          matchedBy = "opening_line"
        } else if (needles.length === 1) {
          const bodyHits = needles.filter((needle) => fuzzyTokenMatch(needle, folded.tokenList)).length
          if (bodyHits / needles.length >= 0.6) {
            score = 42
            matchedBy = "full_text"
          }
        }
      }
      if (!score) {
        if (exactHits && coverage >= 0.6) {
          score = 12 * coverage
          matchedBy = "full_text"
        } else {
          return null
        }
      }
    }
  } else {
    return null
  }
  return {
    number: row.n,
    title: row.t,
    firstLine: row.o,
    snippet:
      matchedBy === "opening_line"
        ? row.o || row.t
        : lyricSnippet(query, tokens, row.o, body) || row.o || row.t,
    score,
    matchedBy,
  }
}

export function isLyricCatalogQuery(value: string) {
  const trimmed = value.trim()
  if (trimmed.length < 2) return false
  if (isCatalogNumberQuery(trimmed)) return false
  if (/^search prabhat samgiita for\s+/i.test(trimmed)) return false
  if (/\?/.test(trimmed)) return false
  if (
    /\b(?:i(?:'m| am)|we are|feel(?:ing)?|help me|recommend|suggest|what|why|how|should i|can you|please)\b/i.test(
      trimmed,
    )
  ) {
    return false
  }
  if (/^(?:songs?|song)\s+(?:for|about|on)\b/i.test(trimmed)) return false
  return true
}

export function confidentLyricHits(hits: LyricSearchHit[]): LyricSearchHit[] {
  return hits.filter((hit) => hit.matchedBy === "opening_line" || hit.score >= 40)
}

export function interpretLyricHits(hits: LyricSearchHit[]): LyricSearchHit[] {
  const confident = confidentLyricHits(hits)
  if (confident.length) return confident
  return hits.filter((hit) => hit.score >= 24)
}

export function searchLyrics(query: string, rows: LyricSearchRow[], limit = 5): LyricSearchHit[] {
  const normalized = normalizeLyricText(query)
  if (normalized.length < 2) return []
  const tokens = normalized.split(" ").filter(Boolean)
  const foldedQuery = foldLyricPhonetic(normalized)
  const foldedTokenByExact = new Map(tokens.map((token) => [token, foldLyricPhonetic(token)]))
  const hits: LyricSearchHit[] = []
  for (const row of rows) {
    const hit = scoreRow(normalized, tokens, row, foldedQuery, foldedTokenByExact)
    if (hit) hits.push(hit)
  }
  hits.sort((left, right) => right.score - left.score || left.number - right.number)
  return hits.slice(0, Math.max(1, limit))
}
